openmc/tallies.py

```diff
@@ -0,0 +1,3204 @@
+                                              self._sum_sq.shape)
+
+            # Indicate that Tally results have been read
+            self._results_read = True
+
+            # Close the HDF5 statepoint file
+            f.close()
+
+        if self.sparse:
+            return np.reshape(self._sum.toarray(), self.shape)
+        else:
+            return self._sum
+
+    @property
+    def sum_sq(self):
+        if not self._sp_filename or self.derived:
+            return None
+
+        if not self._results_read:
+            # Force reading of sum and sum_sq
+            self.sum
+
+        if self.sparse:
+            return np.reshape(self._sum_sq.toarray(), self.shape)
+        else:
+            return self._sum_sq
+
+    @property
+    def mean(self):
+        if self._mean is None:
+            if not self._sp_filename:
+                return None
+
+            self._mean = self.sum / self.num_realizations
+
+            # Convert NumPy array to SciPy sparse LIL matrix
+            if self.sparse:
+                self._mean = sps.lil_matrix(self._mean.flatten(),
+                                            self._mean.shape)
+
+        if self.sparse:
+            return np.reshape(self._mean.toarray(), self.shape)
+        else:
+            return self._mean
+
+    @property
+    def std_dev(self):
+        if self._std_dev is None:
+            if not self._sp_filename:
+                return None
+
+            n = self.num_realizations
+            nonzero = np.abs(self.mean) > 0
+            self._std_dev = np.zeros_like(self.mean)
+            self._std_dev[nonzero] = np.sqrt((self.sum_sq[nonzero]/n -
+                                              self.mean[nonzero]**2)/(n - 1))
+
+            # Convert NumPy array to SciPy sparse LIL matrix
+            if self.sparse:
+                self._std_dev = sps.lil_matrix(self._std_dev.flatten(),
+                                               self._std_dev.shape)
+
+            self.with_batch_statistics = True
+
+        if self.sparse:
+            return np.reshape(self._std_dev.toarray(), self.shape)
+        else:
+            return self._std_dev
+
+    @property
+    def with_batch_statistics(self):
+        return self._with_batch_statistics
+
+    @property
+    def derived(self):
+        return self._derived
+
+    @property
+    def derivative(self):
+        return self._derivative
+
+    @property
+    def sparse(self):
+        return self._sparse
+
+    @estimator.setter
+    def estimator(self, estimator):
+        cv.check_value('estimator', estimator, ESTIMATOR_TYPES)
+        self._estimator = estimator
+
+    @triggers.setter
+    def triggers(self, triggers):
+        cv.check_type('tally triggers', triggers, MutableSequence)
+        self._triggers = cv.CheckedList(openmc.Trigger, 'tally triggers',
+                                        triggers)
+
+    @name.setter
+    def name(self, name):
+        if name is not None:
```

Comment:
we could have a `cv.check_type_or_none` function. We do the `if not None` check everywhere